Bugfix/api view not loading 特殊天

deligence-dharmendra opened the pull request:
Closes #3101

## Changes

- add new node package 'limax'
- chnages in api autoform js
- remove file slugs.js for api name
- remove organization slugs.js
- add new api slug migration
- add new organization slug migration
- add new migration for manage duplicate organization name

## Developer checklist
This checklist is to be completed by the PR developer:

- [ ] Alternative solutions were compared/discussed before writing code
    - [ ] trade-offs with this solution are considered acceptable
- [ ] Code in this PR adheres to the [project styleguide](CONTRIBUTING.md)
- [ ] This pull request does not decrease project test coverage
- [ ] If the code changes existing database collection(s), migration has been written
- [ ] If UI texts are added or changed, all texts are internationalized

## Reviewer checklist
Reviewed by: @marla-singer @matleppa 

This list is to be completed by the pull request reviewer:
- [ ] Code works as described/expected
- [ ] Code seems to be error free
    - [ ] no browser console errors visible
    - [ ] no server console errors visible
    - [ ] passes CI build
- [ ] Code is written in a way that promotes maintainability
    - [ ] easy to understand
    - [ ] well organized
    - [ ] follows project coding standards and conventions
    - [ ] well documented

matleppa commented:
### Good start
When an API having name with non-ascii alphabet is created, this seems to work fine. A converted slug is formed.

### However, 
Same functionality should apply when changing API name. A slug reflecting changed API name should be formed, same way as now happens with API names of ascii characters.

#### What happens now with non-ascii chars?
- add ascii part to non-ascii API name
  - slug is formed of only ascii part
    - should be formed based on complete name
- add non-ascii characters to API name previously with ascii characters
  - new slug is formed based on Roman character part of name (old: slug-for-romans, new: slug-for-romans-1)
    - should be formed based on complete name
- add non-ascii chars to name previously with non-ascii
  - the slug remains the same as earlier
    - should be formed based on complete name
  - exception in server log
```
Exception from sub apiComposite id QczwsBNFBkipA4gj7 Error: Match error: Expected string, got null
    at exports.check (packages\check.js:55:15)
    at [object Object].<anonymous> (apinf_packages/apis/collection/server/publications.js:57:3)
    at [object Object].prepareOptions (packages/reywood:publish-composite/lib/publish_composite.js:37:43)
    at [object Object].publish (packages/reywood:publish-composite/lib/publish_composite.js:12:48)
    at packages\matb33_collection-hooks.js:307:21
    at [object Object].EVp.withValue (packages\meteor.js:1135:15)
    at [object Object]._handler (packages\matb33_collection-hooks.js:306:28)
    at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1768:12)
    at packages/ddp-server/livedata_server.js:1043:15
    at [object Object].EVp.withValue (packages\meteor.js:1135:15)
    at [object Object]._.extend._runHandler (packages/ddp-server/livedata_server.js:1041:51)
    at [object Object]._.extend._startSubscription (packages/ddp-server/livedata_server.js:859:9)
    at [object Object]._.extend.protocol_handlers.sub (packages/ddp-server/livedata_server.js:625:12)
    at packages/ddp-server/livedata_server.js:559:43
Sanitized and reported to the client as: Match failed [400]

```
 - should be no exception

#### Exceptions also in other case 
Similar kind of exceptions is also generated 
1 open an API
2 select Add API (in main menu)

Here the probable cause may be missing check whether slug was gotten (around line 31 in apinf_packages\apis\client\profile\view.js)

marla-singer commented:
@matleppa @deligence-dharmendra I put my two cents. 
The current solution will fix the future cases - if a user **will come** to a site and **will add** an API with no standard alphabet name. 

How does the implemented code fix the existing cases in apinf.io site? I meant does it need to add migration step to update existing slug parameters?

matleppa commented:
@marla-singer @deligence-dharmendra 
Good point, the need of migration.
At the time in apinf.io there are two APIs with non-ascii chars in name: 
- openAPI测试 (slug: openapi)
  - although to slug is not formed of complete API name, it is unique and the API card can be opened. When also functionality of taken into account the non-ascii chars, when API name is changed and new slug is formed, is implemented and taken in use, this (or these) APIs' slug will be better in case the name is changed. 
-  特殊天 (no slug)
  - this API card can not be opened via UI, because of missing slug. 
  
So I think that the migration could go through all APIs and only in case the slug is **missing**, it forms one (of course checking, that the formed slug is unique).
@marla-singer Do you agree or do you find something missing?

deligence-dharmendra commented:
Hi @matleppa 
what i'm getting point from you suggestion, if I'm not wrong.
1)eg:
 **API Name**     =>  **slug**
Activation and Configuration API  => activation-and-configuration-api
24Rent-yhteiskäyttöautot => 24rent-yhteiskaeyttoeautot
apinf要实现 => apinf-yao-shi-xian
特殊天 => te-shu-tian
要实现 Speaker => yao-shi-xian-speaker
 
Please advise

matleppa commented:
@deligence-dharmendra Yes, same functionality concerning non-ascii char handling is needed both in creation on an API and changing of name of an existing API.

And also during migration.

deligence-dharmendra commented:
Hi @matleppa @marla-singer 
I have implemented the functionality for creation and updation for ascii and non-ascii API name.
and it is working fine for me as per my understanding.But i'm getting issue with file apis/collection/slugs.js. 
Can i comment all code in this file.

Please advise

marla-singer commented:
> Can i comment all code in this file.

Comment this file. If everything works fine without the standart slug and with "limax" packages then
1. Remove this file
2. Migrate the new slug implementation

Test the cases:
1. A user creates an API
2. A user updates the API

**Expected result**:
1. The name is updated successfully
2. The slug is updated successfully as well

Also the application supports slug for organizations. If the new slug implementation works well then migrate for organizaiont as well

deligence-dharmendra commented:
Hi @marla-singer @matleppa 
Please review the latest changes . If everything on the right way,  then i will implement same functionality for the organization name.

matleppa commented:
@deligence-dharmendra It seems, that we are going to right direction, however, I was able to find one problem.

API names has to be unique, and also slugs need to be unique.
Because the non-ascii chars in API name are mapped to ascii chars in different ways, it can happen, that the formed slug is not unique. One of those chars is 'å', which is mapped 'a'.
Previously this is avoided by adding a postfix into formed slug ( e.g. -1).

In nightly.apinf.io there is an example of old functionality:

**API name | slug**
AAA new API | aaa-new-api
ÅÅÅ new API | aaa-new-api-1

However there is a problem, when I'm using this branch and trying to create an API giving similar kind of names 
**API name | slug**
api | api
åpi |

When I press button [Add API], instead of creating API card and forming slug such as (api-1), nothing happens.

Same problem when changing API name.

matleppa commented on apinf_packages/apis/server/methods/apiBackends.js:
```diff
@@ -27,4 +30,65 @@ Meteor.methods({
+      // Make sure Id is a string
+      check(apiId, String);
+      const api = Apis.findOne(apiId);
+      apiName = api.name;
```
Check if api exists.

matleppa commented on apinf_packages/apis/server/methods/apiBackends.js:
```diff
@@ -27,4 +30,65 @@ Meteor.methods({
+    const existingApi = Apis.findOne(
+      { 'friendlySlugs.slug.base':slug },
+      { sort: { created_at: -1 }},
+      { limit: 1 }
```
What is purpose of this parameter limit?

matleppa commented on apinf_packages/apis/server/methods/apiBackends.js:
```diff
@@ -27,4 +30,65 @@ Meteor.methods({
+    // Transliterates non-Latin scripts
+    const slug = slugs(apiName, { tone: false });
+
+    // Look for existing duplicate slug
```
// Look for existing duplicate slug beginning of the newest one

marla-singer commented on apinf_packages/core/migrations/server/6-add-slug-for-apis.js:
```diff
@@ -7,7 +7,13 @@ Migrations.add({
   up () {
     Apis.find({ slug: { $exists: false } }).forEach((api) => {
-      Apis.update(api._id);
+      if (api.name) {
```
Never change the old steps of migration. There needs to save persistence instructions. In this case, the different APInf instances may have different behaviour of migration. 

**Create a new migration step**

deligence-dharmendra commented:
Hi @marla-singer @matleppa 
1)I have created a new migration file:"12-add-slug-for-existing-apis" for slug api, and pushed changes to branch. But at the time of travis ci build, I'm getting error and not aware about those error.
**error like**
00 login
    1) "before all" hook
  0 passing (10s)
  1 failing
  1) 00 login "before all" hook:
     Uncaught unknown error: cannot determine loading status
    from unknown error: missing or invalid 'entry.level'

2)I have a doubt about migration version 6. I have implemented slug using new way and new package 'limax' and removed the existing file api slug.js. So it may be create issue at the time migration version 6  will call.

Please advise.

marla-singer commented:
> So it may be create issue at the time migration version 6 will call.

@deligence-dharmendra Test it. But I don't see any dependences on the slug library in the 6 migration step

marla-singer commented:
@deligence-dharmendra The PR contains the alien commits and it is hard to review.

1. make `git rebase develop` - it will place your commits on the current branch  top
2. Fix a conflict if it would be
3. make `git push -f`

marla-singer commented on apinf_packages/organizations/client/profile/settings/general/autoform.js:
```diff
@@ -11,11 +11,30 @@ import { sAlert } from 'meteor/juliancwirko:s-alert';
+      // Getting name from this.udateDoc
+      const orgName = this.updateDoc.$set.name;
+      if (orgName) {
+        Meteor.call('updateOrgSlug', null, orgName, (error, slug) => {
```
Use the full name. 
org = organization
orgName = organizationName
updateOrgSlug = updateOrganizationSlug
and the same stuff

**Remake all short names**

marla-singer replied:
And after updating the function, call it as 
```js
Meteor.call('updateOrganizationSlug', { name: organizationName }, () => { .. });
```

marla-singer commented on apinf_packages/organizations/server/methods.js:
```diff
@@ -175,4 +178,79 @@ Meteor.methods({
 
     return userIsAdmin || userIsManager;
   },
+  updateOrgSlug (organizationId = null, name = null) {
```
Don't two parameters for function "updateOrgSlug" that are not both used at one time. I strongly recommend to use only one parameter - `query` that is the Object and will be `{ _id: organizationId }` or `{ name: organizationName }`.

Use it as
```js
const organization = Organizations.findOne(query);
```

marla-singer commented on apinf_packages/organizations/client/add/autoform.js:
```diff
@@ -41,14 +41,23 @@ AutoForm.hooks({
-        // Redirect to newly added organization
-        FlowRouter.go('organizationProfile', { slug });
+        if (organizationId) {
+          Meteor.call('updateOrgSlug', organizationId, null, (error, result) => {
```
And after updating the function, call it as 
```js
Meteor.call('updateOrganizationSlug', { _id: organizationId }, () => { .. });
```

marla-singer replied:
And why do we need update slug if the organization was added? I think it is more logical to make if formType is update

matleppa commented:
It was a good catch, that also Organization needs correction in forming the slug.

There is one matter, which need to be taken into account, because there is difference between APIs and Organizations.

API name must be unique, but Organization name needs not to be unique.

I tested the Organization slug functionality:
- create an Organization (name with non-ascii chars
  - creation OK, a translittered slug was formed 
- create second Organization **using same name**
  - new Organization was created, but without a slug. 
  - however, the slug of **first** Organization was changed (added index number)
- create third Organization **using same name**
  - new Organization was created, but without a slug. 
  - however, the slug of **first** Organization was changed (increased index number)

So a checking is needed, that slug of correct Organization is formed.

Nazarah commented:
@matleppa apart from this PR, do you recommend that organizations have unique name?
That would be justifiable.
because if you consider normal cases like creating a new gmail ID, google checks that if a given ID already exists or taken and then it suggests an alternative one.

matleppa commented:
@Nazarah Generally, to me it would sound logical, that also Organization names were unique. 

I am not familiar with previous decisions, what might have been the reason not to have unique Organization names.

preriasusi commented:
Would it make sense to handle organisation changes in another issue? Or the issue of organisations name uniqueness in another issue? Feels too much for one.

deligence-dharmendra commented:
Hi @preriasusi 
Do I need to work on same branch for organization slug or I need to create new branch for it.
Second thing I want to know Organization name needs to be unique or not. I think it should be unique. 

Please advise

deligence-dharmendra commented:
Hi @preriasusi @marla-singer @matleppa @Nazarah 
Organization name needs to be unique or not?. I think it should be unique.

Please advise.

matleppa commented:
I agree with unique Organization names.
Of course it means, that a migration is once again needed to make sure, that all already existing Organization names also are unique.

deligence-dharmendra commented:
Hi @matleppa 
Can you suggest me, what step I need to do if organization duplicate name exists, I mean what suffix or prefix need to add.

Please advise

matleppa commented:
I'd say that during migration the found duplicated names are separated with suffix.
-X, where X is increasing number.

Let's say there are three Organizations named "Fine Organization"

After migration they were:
Fine Organization
Fine Organization-1
Fine Organization-2

deligence-dharmendra commented:
Hi @matleppa 
Below conversion made by Limax package, this scenario happened In only some language case like Chinese.

**特殊天-1 --> te-shu-tian1**

I have posted this issue on Limax GitHub repository. https://github.com/lovell/limax/issues/28
Please advise

matleppa commented:
@deligence-dharmendra About the Limax issue: OK, fine. So the fix will probably come at some point. 
And this is the best we can do now.

marla-singer commented on apinf_packages/apis/server/methods/apiBackends.js:
```diff
@@ -27,11 +30,67 @@ Meteor.methods({
+    check(query, Object);
+    const api = Apis.findOne(query);
+    if (!api) {
+      return '';
```
Bad choose. If `api` doesn't exists with provided parameter then 

```js
throw new Meteor.Error(`The API doesn't exists with parameter ${query}`);
```

**Remake it**

marla-singer commented on apinf_packages/organizations/server/methods.js:
```diff
@@ -175,4 +178,68 @@ Meteor.methods({
+    const organization = Organizations.findOne(query);
+
+    if (!organization) {
+      return '';
```
Bad choose. If `organization` doesn't exist with provided parameter then 

```js
throw new Meteor.Error(`The Orgniazaiont doesn't exist with parameter ${query}`);
```

**Remake it**

marla-singer commented on apinf_packages/organizations/client/profile/settings/general/autoform.js:
```diff
@@ -3,19 +3,42 @@ This file is covered by the EUPL license.
+        });
+      } else {
+        // Otherwise Redirect to Organization Catalog
+        FlowRouter.go('organizations');
```
The same stuff is here

If organizationName doesn't exist then **don't redirect to Organization Catalog**. Because a user can edit, for example, only Description field and in this case:
1. organizationName parameter doesn't exist
2. The page redirects to Catalog that is a wrong behavior because It expects to not change a page

**Remove this condition**

marla-singer commented on apinf_packages/apis/client/profile/settings/details/autoform.js:
```diff
@@ -11,14 +14,32 @@ import { FlowRouter } from 'meteor/kadira:flow-router';
+            FlowRouter.go('apiCatalog');
+          } else if (slug && slug !== '') {
+            // Redirect to modified api
+            FlowRouter.go('viewApi', { slug });
```
Simplify this code.

After improving `updateApisBySlug` function will return the `slug` value or throw error then

**Pseudocode**
1. If `error` then display a message and redirect to Catalog
2. If `slug` then redirect to Profile with new slug

**Remake**

marla-singer commented on apinf_packages/organizations/client/profile/settings/general/autoform.js:
```diff
@@ -3,19 +3,42 @@ This file is covered by the EUPL license.
+          if (error) {
+            // Show message
+            sAlert.error(error);
+          } else if (slug && slug !== '') {
```
The same stuff is here

Simplify this code. After improving `updateOrganizationBySlug ` function will return the `slug` value or throw error then

**Pseudocode**
1. If `error` then display a message and redirect to Catalog
2. If `slug` then display a message and redirect to Profile with new slug

**Remake**

marla-singer commented on apinf_packages/apis/client/add/autoform.js:
```diff
@@ -29,15 +30,24 @@ AutoForm.hooks({
+            // Show message
+            sAlert.error(error);
+            FlowRouter.go('apiCatalog');
+          } else if (result && result !== '') {
```
1. **Rename "result" variable to "slug"
2. Simplify this code. After improving `updateApisBySlug ` function will return the `slug` value or throw error then

**Pseudocode**
1. If `error` then display a message and redirect to Catalog
2. If `slug` then redirect to Profile with new slug

**Remake**

marla-singer replied:
Actually, why do we need Update slug if a user just added an API?

deligence-dharmendra replied:
Hi @marla-singer 
Actually, API does not have the slug value in the form field while inserting. So we need to add slug manually by calling this meteor call.

marla-singer commented on apinf_packages/core/migrations/server/14-add-slug-for-existing-apis.js:
```diff
@@ -0,0 +1,27 @@
+import Apis from '/apinf_packages/apis/collection';
+
+Migrations.add({
+  version: 14,
```
Up your version of Migration steps to one

marla-singer commented on apinf_packages/core/migrations/server/14-add-slug-for-existing-apis.js:
```diff
@@ -0,0 +1,27 @@
+      if (api.name) {
+        Meteor.call('updateApisBySlug', { name: api.name }, (error) => {
+          if (error) {
+            //  Show error messag
```
Why do you do it if no action happens?

1. If you want to sync call then 
```js
const call = Meteor.call('updateApisBySlug', { name: api.name });
```
2. If you don't need a sync call then
```js
Meteor.call('updateApisBySlug', { name: api.name });
```

**Remake it**

marla-singer commented on apinf_packages/core/migrations/server/16-add-slug-for-existing-organizations.js:
```diff
@@ -0,0 +1,27 @@
+      if (organization.name) {
+        Meteor.call('updateOrganizationBySlug', { name: organization.name }, (error) => {
+          if (error) {
+            //  Show error messag
```
Why do you do it if no action happens?

1. If you want to sync call then 
```js
const call = Meteor.call('updateOrganizationBySlug', { name: organization.name });
```
2. If you don't need a sync call then
```js
Meteor.call('updateOrganizationBySlug', { name: organization.name });
```

**Remake it**

marla-singer commented:
@matleppa Will merge it when a new release will be created
